File: hs/Day12.hs
#! runghc

module Main where

data Op = N | S | W | E | L | R | F deriving Show

data Instruction = Instruction { op :: Op, arg :: Int } deriving Show

data Point = Point { px :: Int, py :: Int } deriving Show

parseInstruction :: String -> Instruction
parseInstruction (x:xs) =
    case x of
        'N' -> Instruction N (read xs)
        'S' -> Instruction S (read xs)
        'W' -> Instruction W (read xs)
        'E' -> Instruction E (read xs)
        'L' -> Instruction L (read xs)
        'R' -> Instruction R (read xs)
        'F' -> Instruction F (read xs)

readInputs :: FilePath -> IO [Instruction]
readInputs path = do
    s <- lines <$> readFile path
    return $ map parseInstruction s

moveNorth :: Point -> Int -> Point
moveNorth (Point x y) arg = Point x (y + arg)

moveSouth :: Point -> Int -> Point
moveSouth (Point x y) arg = Point x (y - arg)

moveWest :: Point -> Int -> Point
moveWest (Point x y) arg = Point (x - arg) y

moveEast :: Point -> Int -> Point
moveEast (Point x y) arg = Point (x + arg) y

rotatePoint :: Point -> Int -> Point
rotatePoint (Point x y) degrees = 
    

    let x' = cos' * x - sin' * y in
    let y' = sin' * x + cos' * y in
            Point x' y'

    where 
        cos' = case degrees `mod` 360 of
                    0 -> 1
                    90 -> 0
                    180 -> negate 1
                    270 -> 0
        sin' = case degrees `mod` 360 of
                    0 -> 0
                    90 -> 1
                    180 -> 0
                    270 -> negate 1

navigate :: Point -> Point -> [Instruction] -> IO Point
navigate ship waypoint [] = return ship
navigate ship waypoint (c:cs) = do

    let ship' = case c of
            Instruction N arg -> do
                moveNorth ship arg
            Instruction S arg -> do
                moveSouth ship arg
            Instruction W arg -> do
                moveWest ship arg
            Instruction E arg -> do
                moveEast ship arg
            Instruction L arg -> do
                ship
            Instruction R arg -> do
                ship
            Instruction F arg -> do
                addManyTimes ship waypoint arg

    let waypoint' = case c of
            Instruction N arg -> do
                waypoint
            Instruction S arg -> do
                waypoint
            Instruction W arg -> do
                waypoint
            Instruction E arg -> do
                waypoint
            Instruction L arg -> do
                rotatePoint waypoint arg
            Instruction R arg -> do
                rotatePoint waypoint (negate arg)
            Instruction F arg -> do
                waypoint

    navigate ship' waypoint' cs

navigate' :: Point -> Point -> [Instruction] -> IO Point
navigate' ship waypoint [] = return ship
navigate' ship waypoint (c:cs) = do

    let ship' = case c of
            Instruction N arg -> do
                ship
            Instruction S arg -> do
                ship
            Instruction W arg -> do
                ship
            Instruction E arg -> do
                ship
            Instruction L arg -> do
                ship
            Instruction R arg -> do
                ship
            Instruction F arg -> do
                addManyTimes ship waypoint arg

    let waypoint' = case c of
            Instruction N arg -> do
                moveNorth waypoint arg
            Instruction S arg -> do
                moveSouth waypoint arg
            Instruction W arg -> do
                moveWest waypoint arg
            Instruction E arg -> do
                moveEast waypoint arg
            Instruction L arg -> do
                rotatePoint waypoint arg
            Instruction R arg -> do
                rotatePoint waypoint (negate arg)
            Instruction F arg -> do
                waypoint

    navigate' ship' waypoint' cs

addManyTimes :: Point -> Point -> Int -> Point
addManyTimes (Point a b) (Point c d) n = Point (a + n * c)  (b + n * d)

size :: Point ->  Int
size (Point x y) = abs x + abs y

main :: IO ()
main = do

    instructions <- readInputs "../inputs/Day12.txt"
    
    point <- navigate (Point 0 0) (Point 1 0) instructions

    print $ size point

    point <- navigate' (Point 0 0) (Point 10 1) instructions

    print $ size point

    return ()

File: py/day12.py
#! python3

from dataclasses import dataclass


@dataclass
class Point:
    x: float
    y: float

    def copy(self):
        return Point(self.x, self.y)

    @property
    def size(self):
        return abs(self.x) + abs(self.y)

    def move(self, dx, dy):
        self.x += dx
        self.y += dy

    def move_north(self, n):
        self.move(0, n)

    def move_south(self, n):
        self.move(0, -n)

    def move_east(self, n):
        self.move(n, 0)

    def move_west(self, n):
        self.move(-n, 0)

    def rotate(self, degrees):

        cos_ = cos(degrees)
        sin_ = sin(degrees)

        new_x = self.x * cos_ - self.y * sin_
        new_y = self.x * sin_ + self.y * cos_

        self.x = new_x
        self.y = new_y


@dataclass
class Instruction:
    op: str
    arg: int


def sin(theta):

    return {
        0: 0,
        90: 1,
        180: 0,
        270: -1
    }[theta % 360]


def cos(theta):

    return {
        0: 1,
        90: 0,
        180: -1,
        270: 0
    }[theta % 360]


def line_to_instruction(line):

    instruction = line[0]
    argument = int(line[1:])

    return Instruction(instruction, argument)


def load_instructions(path):

    return [
        line_to_instruction(line)
        for line in open(path, "r").readlines()
    ]

def run_instructions1(location: Point, waypoint: Point, instructions: Instruction):

    for instruction in instructions:

        op = instruction.op
        arg = instruction.arg

        if op == "N":
            location.move_north(arg)
        elif op == "S":
            location.move_south(arg)
        elif op == "W":
            location.move_west(arg)
        elif op == "E":
            location.move_east(arg)
        elif op == "L":
            waypoint.rotate(arg)
        elif op == "R":
            waypoint.rotate(-arg)
        elif op == "F":
            location.x += waypoint.x * arg
            location.y += waypoint.y * arg

    return (location, waypoint)


def run_instructions2(location, waypoint, instructions):

    for instruction in instructions:

        op = instruction.op
        arg = instruction.arg

        if op == "N":
            waypoint.move_north(arg)
        elif op == "S":
            waypoint.move_south(arg)
        elif op == "W":
            waypoint.move_west(arg)
        elif op == "E":
            waypoint.move_east(arg)
        elif op == "L":
            waypoint.rotate(arg)
        elif op == "R":
            waypoint.rotate(-arg)

        elif op == "F":

            location.x += waypoint.x * arg
            location.y += waypoint.y * arg

    return (location, waypoint)


instructions = load_instructions("../inputs/Day12.txt")

(location, direction) = run_instructions1(Point(0, 0), Point(1, 0), instructions)

print(location.size)

(location, waypoint) = run_instructions2(
    Point(0, 0), Point(10, 1), instructions)

print(location.size)
